design/one-pager-package-runtime-config.md

```diff
@@ -0,0 +1,222 @@
+   configuration of the Deployment (etc) Crossplane creates. ([#3601])
+2. In some cases, folks don't want Crossplane to create a Kubernetes Deployment
+   at all. They want to run a Provider some other way. ([#2671])
+
+Soon [Composition Functions][functions-beta-design] will also be deployed by the
+package manager as Kubernetes Deployments. We expect they'll have similar
+requirements. Since Functions aren't Kubernetes controllers, in this document
+I'll refer to the long-running processes some packages need to deploy as
+'package runtimes'.
+
+Crossplane has a v1alpha1 ControllerConfig type that addresses the first issue
+for Providers. It has been marked deprecated, to be removed if and when we find
+a suitable replacement. We deprecated ControllerConfig because:
+
+* It was growing piecemeal to support templatizing an entire Deployment.
+* We think in some rare cases package runtimes won't use Deployments.
+
+It's worth noting that the desire to deploy a package runtime as anything other
+than a Kubernetes Deployment in the same cluster where Crossplane is running is
+_quite rare_. To my knowledge only Upbound currently has this requirement.
+
+## Goals
+
+The goals of this design are to:
+
+* Give folks full control over package runtime Deployments.
+* Make it possible to run package runtimes as something other than a Deployment.
+
+## Proposal
+
+I propose a new flag to Crossplane: `--package-runtime`. This flag would have
+two possible values (at least to begin with):
+
+* `--package-runtime=Deployment` (default) - Create a Kubernetes deployment.
+* `--package-runtime=External` - Do nothing, defer to an external controller.
+
+When running in Deployment mode the package manager will function as it does
+today. It will create a Kubernetes Deployment for each package runtime, plus any
+additional supporting configuration such as a ServiceAccount, etc.
+
+When running in External mode the package manager won't create a package runtime
+at all. It will create a revision (e.g. a ProviderRevision) and deliver the
+package's payload (e.g. a Provider's CustomResourceDefinitions), but do nothing
+else. In External mode it's expected that an external controller will take care
+of reconciling the relevant package revision to deploy a package runtime however
+it sees fit.
+
+I also propose we replace ControllerConfig with a new DeploymentRuntimeConfig
+type in the pkg.crossplane.io API group. This type is used to configure the
+package runtime when the package manager is running with
+`--package-runtime=Deployment`.
+
+A DeploymentRuntimeConfig is referenced from any package that uses a runtime,
+such as a Provider. For example:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1
+kind: Provider
+metadata:
+  name: provider-example
+spec:
+  package: xpkg.upbound.io/crossplane-contrib/provider-example:v1.0.0
+  runtimeConfigRef:
+    apiVersion: pkg.crossplane.io/v1
+    kind: DeploymentRuntimeConfig
```

Comment:
```suggestion
    apiVersion: runtime.pkg.crossplane.io/v1
    kind: DeploymentConfig
```

just as another idea :)

Reply:
Haha I almost proposed exactly this, but was thinking perhaps a "flatter" API group namespace would be better. I could go either way on this - curious to hear whether folks have a preference.